src/Nonlinear/operators.jl

```diff
@@ -0,0 +1,287 @@
+        push!(registry.registered_univariate_operators, operator)
+        registry.univariate_operator_to_id[op] =
+            length(registry.univariate_operators)
+    else
+        operator = MultivariateOperator{nargs}(f...)
+        push!(registry.multivariate_operators, op)
+        push!(registry.registered_multivariate_operators, operator)
+        registry.multivariate_operator_to_id[op] =
+            length(registry.multivariate_operators)
+    end
+    return
+end
+
+function eval_univariate_function(
+    registry::OperatorRegistry,
+    op::Symbol,
+    x::T,
+)::T where {T}
+    id = registry.univariate_operator_to_id[op]
+    if id <= registry.univariate_user_operator_start
+        f, _ = _eval_univariate(id, x)
+        return f
+    else
+        offset = id - registry.univariate_user_operator_start
+        operator = registry.registered_univariate_operators[offset]
+        return operator.f(x)
+    end
+end
+
+function eval_univariate_gradient(
+    registry::OperatorRegistry,
+    op::Symbol,
+    x::T,
+) where {T}
+    id = registry.univariate_operator_to_id[op]
+    if id <= registry.univariate_user_operator_start
+        _, f′ = _eval_univariate(id, x)
+        return f′
+    else
+        offset = id - registry.univariate_user_operator_start
+        operator = registry.registered_univariate_operators[offset]
+        return operator.f′(x)
+    end
+end
+
+function eval_univariate_hessian(
+    registry::OperatorRegistry,
+    op::Symbol,
+    x::T,
+) where {T}
+    id = registry.univariate_operator_to_id[op]
+    if id <= registry.univariate_user_operator_start
+        return _eval_univariate_2nd_deriv(id, x)
+    else
+        offset = id - registry.univariate_user_operator_start
+        operator = registry.registered_univariate_operators[offset]
+        return operator.f′′(x)
+    end
+end
+
+function eval_multivariate_function(
+    registry::OperatorRegistry,
+    op::Symbol,
+    x::AbstractVector{T},
+)::T where {T}
+    if op == :+
+        return +(x...)
+    elseif op == :-
+        return -(x...)
+    elseif op == :*
+        return *(x...)
+    elseif op == :^
+        @assert length(x) == 2
+        return x[1]^x[2]
+    elseif op == :/
+        @assert length(x) == 2
+        return x[1] / x[2]
+    elseif op == :ifelse
+        @assert length(x) == 3
+        return ifelse(Bool(x[1]), x[2], x[3])
+    end
+    id = registry.multivariate_operator_to_id[op]
+    offset = id - registry.multivariate_user_operator_start
+    operator = registry.registered_multivariate_operators[offset]
+    @assert length(x) == operator.N
+    return operator.f(x)
+end
+
+function eval_multivariate_gradient(
+    registry::OperatorRegistry,
+    op::Symbol,
+    g::AbstractVector{T},
+    x::AbstractVector{T},
+) where {T}
+    if op == :+
+        fill!(g, 1.0)
+    elseif op == :-
+        fill!(g, -1.0)
+        g[1] = 1.0
+    elseif op == :*
+        total = *(x...)
+        for i in 1:length(x)
```

Comment:
This rule doesn't work if total == 0. See https://github.com/jump-dev/JuMP.jl/blob/8ecd560daa56ab15626da8f9f48e94d68562be3a/src/_Derivatives/forward.jl#L94-L116.

Reply:
Yeah there's going to be quite a few of these issues. I tried to do an in-place refactor in https://github.com/jump-dev/JuMP.jl/pull/2937 (+ a lot of uncommitted local changes), but it quickly got incredibly messy as there are too many linkages between `parse_nlp.jl`, `nlp.jl`, and `_Derivatives`. 

My plan is to be in a rush to merge this branch, and to keep all of the existing tests to catch any differences. But it's going to need some careful evaluation before we consider replacing the current AD without breakage.